### PR TITLE
Fix voucher offers API endpoint

### PR DESCRIPTION
**Pull request by marjev**
@mjfrey Here's a PR for the voucher offers API endpoint fix. For a single course coupon, Course Catalog won't be called. Instead, course info will be fetched from LMS API.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=808&repo=edx%2Fecommerce

**Comment by mjfrey** [comment]
@mattdrayer

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 1]
```diff
@@ -80,36 +88,79 @@ def get_offers(self, products, request, voucher):
...
-            q=benefit.range.catalog_query
-        )['results']
 
-        course_ids = [product.course_id for product in products]
-        courses = Course.objects.filter(id__in=course_ids)
-        contains_verified_course = next((False for course in courses if course.type != 'verified'), True)
+        if catalog_query:
+            query_results = request.site.siteconfiguration.course_catalog_api_client.course_runs.get(
+                q=catalog_query
+            )['results']
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/views/vouchers.py]: What if an error is returned by the server during the GET operation?

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 2]
```diff
@@ -80,36 +88,79 @@ def get_offers(self, products, request, voucher):
...
+                        benefit=benefit,
+                        course=course,
+                        course_info=course_catalog_data,
+                        is_verified=contains_verified_course,
+                        offers=offers,
+                        stock_record=stock_record,
+                        voucher=voucher
+                    )
+        else:
+            product = products[0]
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/views/vouchers.py]: If there are no products this will raise an IndexError

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 3]
```diff
@@ -80,36 +88,79 @@ def get_offers(self, products, request, voucher):
...
+            course_id = product.course_id
+            stock_record = StockRecord.objects.get(product__id=product.id)
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/views/vouchers.py:100]: Do we need to defend against StockRecord.DoesNotExist here?

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 4]
```diff
@@ -80,36 +88,79 @@ def get_offers(self, products, request, voucher):
...
+
+            course_info = get_course_info_from_lms(course_id)
+            course = Course.objects.get(id=course_id)
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/views/vouchers.py]: Do we need to defend against Course.DoesNotExist here?

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 5]
```diff
@@ -80,36 +88,79 @@ def get_offers(self, products, request, voucher):
...
+                    course=course,
+                    course_info=course_info,
+                    is_verified=(course.type == 'verified'),
+                    offers=offers,
+                    stock_record=stock_record,
+                    voucher=voucher
+                )
         return offers
+
+    def get_course_offer_data(self, benefit, course, course_info, is_verified, offers, stock_record, voucher):
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/views/vouchers.py]: This method is named 'get' but it's actually modifying the `offers` container -- imho the method should have a different name.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 6]
```diff
@@ -80,36 +88,79 @@ def get_offers(self, products, request, voucher):
...
-                    'course_start_date': course_catalog_data['start'],
-                    'id': course.id,
-                    'image_url': course_catalog_data['image']['src'],
-                    'organization': CourseKey.from_string(course.id).org,
-                    'seat_type': course.type,
-                    'stockrecords': serializers.StockRecordSerializer(stock_record).data,
-                    'title': course.name,
-                    'voucher_end_date': voucher.end_datetime,
-                })
+                course = courses.get(id=course_id)
```
Comment (mattdrayer) [inline ecommerce/extensions/api/v2/views/vouchers.py]: Why are we doing courses.get() here, but Courses.objects.get() on line 125?  Either way we are getting a course object -- seems like we could simplify this by consolidating the work up to the top of the products loop?

**Comment by mattdrayer** [comment]
@marjev @mjfrey -- made a pass through -- let me know when you're ready for another look.

**Comment by mjfrey** [comment]
👍

File: ecommerce/extensions/api/v2/tests/views/test_vouchers.py [hunk 7]
```diff
@@ -145,3 +204,28 @@ def test_get_offers(self):
...
+        voucher, __ = prepare_voucher(_range=new_range, benefit_value=10)
+        voucher, products = get_voucher_and_products_from_code(voucher.code)
+        request = self.prepare_offers_listing_request(voucher.code)
+        self.mock_course_api_response(course=course)
+        offers = VoucherViewSet().get_offers(products=products, request=request, voucher=voucher)
+
+        with mock.patch('ecommerce.extensions.api.v2.views.vouchers.Course.objects.get') as mock_get:
+            mock_get.side_effect = Course.DoesNotExist
+            offers = VoucherViewSet()._get_offers_for_product(products[0], voucher)  # pylint:disable=protected-access
+            self.assertEqual(len(offers), 0)
```
Comment (marjev) [inline ecommerce/extensions/api/v2/tests/views/test_vouchers.py:220]: Don't mock this, delete the course. This is the edge case, which would indicate a deeper problem. Do we want to present the end user with empty page when no Course is found in ecommerce?

File: ecommerce/extensions/api/v2/tests/views/test_vouchers.py [hunk 8]
```diff
@@ -145,3 +204,28 @@ def test_get_offers(self):
...
+
+        with mock.patch('ecommerce.extensions.api.v2.views.vouchers.StockRecord.objects.get') as mock_get:
+            mock_get.side_effect = StockRecord.DoesNotExist
+            offers = VoucherViewSet()._get_offers_for_product(products[0], voucher)  # pylint:disable=protected-access
+            self.assertEqual(len(offers), 0)
```
Comment (marjev) [inline ecommerce/extensions/api/v2/tests/views/test_vouchers.py:225]: Don't mock this, delete the stock record. This is the edge case, which would indicate a deeper problem. Do we want to present the end user with empty page when no Stock Record is found in ecommerce?

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 9]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+                offers = self.get_offers(products, request, voucher)
+            except (ConnectionError, SlumberBaseException, Timeout):
+                logger.error('Could not get course information.')
+                return Response(status=status.HTTP_400_BAD_REQUEST)
             cache.set(cache_hash, offers, settings.COURSES_API_CACHE_TIMEOUT)
 
         page = self.paginate_queryset(offers)
         return self.get_paginated_response(page)
 
+    def _get_offers_for_product(self, product, voucher, course_info=None, offers=None):
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:89]: Why is `offers` an optional parameter when the whole point of this method is appending a dictionary value to it?

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 10]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+        """
+        Retrieves product-specific offers based on the course and benefit information
+        """
+        course_id = product.course_id
+        try:
+            course = Course.objects.get(id=course_id)
+        except Course.DoesNotExist:
+            return offers if offers else []
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:97]: In this case, we have a deeper problem in the database. If Course is not found, we should let the user know what type of error occurred. This way, an empty page would be a result and I'm not sure we want that.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 11]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+
+        try:
+            stock_record = StockRecord.objects.get(product__id=product.id)
+        except StockRecord.DoesNotExist:
+            return offers if offers else []
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:102]: Same comment as above.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 12]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
         page = self.paginate_queryset(offers)
         return self.get_paginated_response(page)
 
+    def _get_offers_for_product(self, product, voucher, course_info=None, offers=None):
+        """
+        Retrieves product-specific offers based on the course and benefit information
+        """
+        course_id = product.course_id
+        try:
+            course = Course.objects.get(id=course_id)
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:95]: Why are we making separate queries to the DB for each course (in case multiple courses are associated with the voucher)? This will definitely slow things down.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 13]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+        """
+        catalog_query = self._get_voucher_catalog_query(voucher)
+        client = request.site.siteconfiguration.course_catalog_api_client
+        try:
+            response = client.course_runs.get(
+                q=catalog_query,
+                page_size=DEFAULT_CATALOG_PAGE_SIZE,
+                limit=DEFAULT_CATALOG_PAGE_SIZE
+            )
+        except (ConnectionError, SlumberBaseException, Timeout):  # pragma: no cover
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:141]: This is redundant. Line 81 catches these exceptions. Same exceptions can happen in line 105.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 14]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+            logger.error('Could not get course run information.')
+            return Response(status=status.HTTP_400_BAD_REQUEST)
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:143]: This Response object will be returned to the `offers` method, but since the Response object can't be paginated, error will occur.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 15]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+                return Response(status=status.HTTP_400_BAD_REQUEST)
             cache.set(cache_hash, offers, settings.COURSES_API_CACHE_TIMEOUT)
 
         page = self.paginate_queryset(offers)
         return self.get_paginated_response(page)
 
+    def _get_offers_for_product(self, product, voucher, course_info=None, offers=None):
+        """
+        Retrieves product-specific offers based on the course and benefit information
+        """
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:92]: Nitpick: Docstring is missing parameters and return value description.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 16]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+                'title': course.name,
+                'voucher_end_date': voucher.end_datetime,
+            })
+
+        return offers
+
+    def _get_offers_for_catalog_query(self, request, products, voucher):
+        """
+        Batch workflow to retrieve product offers for a given catalog query
+        """
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:132]: Nitpick: Docstring is missing parameters and return value description.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 17]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+        for product in products:
+            course_id = product.course_id
+            course_info = next((result for result in query_results if result['key'] == course_id), None)
+            offers = self._get_offers_for_product(product, voucher, course_info, offers)
+        return offers
+
+    def _get_voucher_benefit(self, voucher):
+        """
+        Retrieves a benefit for the first offer from the specified voucher
+        """
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:159]: Nitpick: Docstring is missing parameters and return value description.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 18]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+        return voucher.offers.first().benefit
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:160]: Since we don't have a voucher model in our project to make this a property in, we should place this somewhere in the Voucher application (probably utils).

If we want to DRY this, it should be available everywhere in the project. This has nothing to do with the Voucher API.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 19]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+
+    def _get_voucher_catalog_query(self, voucher):
+        """
+        Retrieves the catalog query, from the range, from the benefit, from the offer, from the specified voucher...
+        """
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:165]: Nitpick: Docstring missing parameters and return value description.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 20]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+        benefit = self._get_voucher_benefit(voucher)
+        return benefit.range.catalog_query
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:167]: Consider placing this somewhere in the Offers application or wherever the Benefit model lives in the Oscar. This has nothing to do with the API endpoint, so it shouldn't be a private method encapsulated here.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 21]
```diff
@@ -90,37 +176,14 @@ def get_offers(self, products, request, voucher):
         Returns:
             List: List of course offers where each offer is represented by a dictionary
         """
-        benefit = voucher.offers.first().benefit
-        offers = []
-        query_results = request.site.siteconfiguration.course_catalog_api_client.course_runs.get(
-            q=benefit.range.catalog_query, page_size=DEFAULT_CATALOG_PAGE_SIZE, limit=DEFAULT_CATALOG_PAGE_SIZE
-        )['results']
+        # If a voucher was not provided, return an empty offer set
+        if not voucher or not products:
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:180]: This is handled by the `offers` method and seems redundant.

File: ecommerce/extensions/api/v2/views/vouchers.py [hunk 22]
```diff
@@ -74,12 +76,96 @@ def offers(self, request):
...
+
+        benefit = self._get_voucher_benefit(voucher)
+
+        if benefit and course_info:
+
+            if offers is None:
+                offers = []
+            offers.append({
+                'benefit': serializers.BenefitSerializer(benefit).data,
+                'contains_verified': (course.type == 'verified'),
```
Comment (marjev) [inline ecommerce/extensions/api/v2/views/vouchers.py:116]: Important: What is the use case for this?

Currently, this causes a bug on the offers page. This definitely doesn't describe the entire offers list but instead sets this flag for one offer only (and that's the only case this will work for, as well as the case when one of the pages consists of the verified seats).
